File: src/packages/control-ui/src/api/sni_segmentation_masquerader.ts
export const SegmentationStrategy = {
  SniBorderSplit: 'SniBorderSplit',
  MidSniSplit: 'MidSniSplit',
  RandomSplit: 'RandomSplit',
} as const;
export type SegmentationStrategy = (typeof SegmentationStrategy)[keyof typeof SegmentationStrategy];

export class SniSegmentationMasquerader {
  public strategy: SegmentationStrategy;
  public minChunkSize: number;
  public maxChunkSize: number;

  constructor(
    strategy: SegmentationStrategy = SegmentationStrategy.MidSniSplit,
    minChunkSize: number = 10,
    maxChunkSize: number = 50,
  ) {
    this.strategy = strategy;
    this.minChunkSize = Math.max(2, minChunkSize);
    this.maxChunkSize = Math.max(minChunkSize + 1, maxChunkSize);
  }

  extractSni(data: Uint8Array): { sni: string; start: number; end: number } | null {
    if (data.length < 43 || data[0] !== 0x16) {
      return null;
    }

    let idx = 43;
    if (idx >= data.length) return null;

    // Skip session ID
    const sessionIdLen = data[idx]!;
    idx += 1 + sessionIdLen;
    if (idx + 2 >= data.length) return null;

    // Skip cipher suites
    const cipherLen = (data[idx]! << 8) | data[idx + 1]!;
    idx += 2 + cipherLen;
    if (idx + 1 >= data.length) return null;

    // Skip compression methods
    const compLen = data[idx]!;
    idx += 1 + compLen;
    if (idx + 2 >= data.length) return null;

    // Extensions length
    const extLen = (data[idx]! << 8) | data[idx + 1]!;
    idx += 2;
    const extEnd = Math.min(idx + extLen, data.length);

    while (idx + 4 <= extEnd) {
      const extType = (data[idx]! << 8) | data[idx + 1]!;
      const extSize = (data[idx + 2]! << 8) | data[idx + 3]!;
      idx += 4;

      if (extType === 0) {
        if (idx + extSize <= extEnd && extSize >= 5) {
          const sniNameLen = (data[idx + 3]! << 8) | data[idx + 4]!;
          const sniStart = idx + 5;
          const sniEnd = sniStart + sniNameLen;
          if (sniEnd <= idx + extSize && sniEnd <= data.length) {
            const sni = new TextDecoder().decode(data.subarray(sniStart, sniEnd));
            return { sni, start: sniStart, end: sniEnd };
          }
        }
      }
      idx += extSize;
    }

    return null;
  }

  segmentStream(data: Uint8Array, seed: number = 42): Uint8Array[] {
    if (data.length === 0) return [];

    const extracted = this.extractSni(data);
    if (extracted) {
      const { start, end } = extracted;
      if (this.strategy === SegmentationStrategy.SniBorderSplit) {
        const chunks: Uint8Array[] = [];
        if (start > 0) chunks.push(data.subarray(0, start));
        chunks.push(data.subarray(start, end));
        if (end < data.length) chunks.push(data.subarray(end));
        return chunks;
      } else if (this.strategy === SegmentationStrategy.MidSniSplit) {
        const mid = start + Math.floor((end - start) / 2);
        return [data.subarray(0, mid), data.subarray(mid)];
      }
    }

    // Pseudo-random split using LCG
    let rngState = seed >>> 0;
    const nextRng = () => {
      rngState = (Math.imul(1664525, rngState) + 1013904223) >>> 0;
      return rngState;
    };

    const chunks: Uint8Array[] = [];
    let curr = 0;
    const range = Math.max(1, this.maxChunkSize - this.minChunkSize);

    while (curr < data.length) {
      const remaining = data.length - curr;
      let step: number;
      if (remaining <= this.minChunkSize) {
        step = remaining;
      } else {
        step = Math.min(this.minChunkSize + (nextRng() % range), remaining);
      }
      chunks.push(data.subarray(curr, curr + step));
      curr += step;
    }

    return chunks;
  }
}

File: labs/mobile/donor-archive/tunnel/SniSegmentationMasquerader.kt
package com.luminet.android.tunnel

import java.nio.ByteBuffer
import java.util.Random

enum class SniSegmentationStrategy {
    SNI_BORDER_SPLIT,
    MID_SNI_SPLIT,
    RANDOM_SPLIT
}

class SniSegmentationMasquerader(
    val strategy: SniSegmentationStrategy,
    val minChunkSize: Int = 8,
    val maxChunkSize: Int = 32
) {
    fun extractSni(data: ByteArray): Triple<String, Int, Int>? {
        if (data.size < 43 || data[0] != 0x16.toByte()) {
            return null
        }

        var idx = 43
        if (idx >= data.size) return null

        val sessionIdLen = data[idx].toInt() and 0xff
        idx += 1 + sessionIdLen
        if (idx + 2 >= data.size) return null

        val cipherLen = ByteBuffer.wrap(data, idx, 2).short.toInt() and 0xffff
        idx += 2 + cipherLen
        if (idx + 1 >= data.size) return null

        val compLen = data[idx].toInt() and 0xff
        idx += 1 + compLen
        if (idx + 2 >= data.size) return null

        val extLen = ByteBuffer.wrap(data, idx, 2).short.toInt() and 0xffff
        idx += 2
        val extEnd = (idx + extLen).coerceAtMost(data.size)

        while (idx + 4 <= extEnd) {
            val extType = ByteBuffer.wrap(data, idx, 2).short.toInt() and 0xffff
            val extSize = ByteBuffer.wrap(data, idx + 2, 2).short.toInt() and 0xffff
            idx += 4

            if (extType == 0 && idx + extSize <= extEnd && extSize >= 5) {
                val sniNameLen = ByteBuffer.wrap(data, idx + 3, 2).short.toInt() and 0xffff
                val sniStart = idx + 5
                val sniEnd = sniStart + sniNameLen
                if (sniEnd <= idx + extSize) {
                    val sni = String(data, sniStart, sniNameLen, Charsets.UTF_8)
                    return Triple(sni, sniStart, sniEnd)
                }
            }
            idx += extSize
        }

        return null
    }

    fun segmentStream(data: ByteArray, seed: Long): List<ByteArray> {
        if (data.isEmpty()) return emptyList()

        val sniInfo = extractSni(data)
        if (sniInfo != null) {
            val (_, start, end) = sniInfo
            when (strategy) {
                SniSegmentationStrategy.SNI_BORDER_SPLIT -> {
                    val chunks = ArrayList<ByteArray>()
                    if (start > 0) chunks.add(data.copyOfRange(0, start))
                    chunks.add(data.copyOfRange(start, end))
                    if (end < data.size) chunks.add(data.copyOfRange(end, data.size))
                    return chunks
                }
                SniSegmentationStrategy.MID_SNI_SPLIT -> {
                    val mid = start + (end - start) / 2
                    return listOf(data.copyOfRange(0, mid), data.copyOfRange(mid, data.size))
                }
                SniSegmentationStrategy.RANDOM_SPLIT -> {
                    // fallthrough
                }
            }
        }

        val rng = Random(seed)
        val chunks = ArrayList<ByteArray>()
        var curr = 0

        while (curr < data.size) {
            val rem = data.size - curr
            val step = if (rem <= minChunkSize) {
                rem
            } else {
                val span = (maxChunkSize - minChunkSize).coerceAtLeast(1)
                (minChunkSize + rng.nextInt(span)).coerceAtMost(rem)
            }
            chunks.add(data.copyOfRange(curr, curr + step))
            curr += step
        }

        return chunks
    }
}
